inference-engine/src/inference_engine/ie_core.cpp

```diff
@@ -506,18 +426,101 @@ class Core::Impl : public ICore {
+        auto parsed = parseDeviceNameIntoConfig(context->getDeviceName(), config);
+        auto plugin = GetCPPPluginByName(parsed._deviceName);
+        bool loadedFromCache = false;
+        ExecutableNetwork res;
+        std::string hash;
+        if (DeviceSupportsImportExport(plugin)) {
+            hash = CalculateNetworkHash(network, parsed._deviceName, plugin, config);
+            try {
+                res = LoadNetworkFromCache(hash, plugin, parsed._config, context, loadedFromCache);
+            } catch (const std::exception &) {
+            }
+        }
+
+        if (!loadedFromCache) {
+            res = LoadNetworkImpl(network, plugin, config, context, hash);
+        }
+        return res;
+    }
+
+    std::string CalculateNetworkHash(const CNNNetwork& network, const std::string& deviceFamily,
+                                     const InferencePlugin& plugin,
+                                     const std::map<std::string, std::string>& config) const {
+
+        // TODO: mnosov: review it
+        auto compileConfig = config;
+        std::map<std::string, Parameter> getMetricConfig;
+
+        // 0. remove DEVICE_ID key
+        auto deviceIt = compileConfig.find(CONFIG_KEY(DEVICE_ID));
+        if (deviceIt != compileConfig.end()) {
+            getMetricConfig[deviceIt->first] = deviceIt->second;
+            compileConfig.erase(deviceIt);
+        }
+
+        // 1. replace it with DEVICE_ARCHITECTURE value
+        std::vector<std::string> supportedMetricKeys =
+            plugin.GetMetric(METRIC_KEY(SUPPORTED_METRICS), getMetricConfig);
+        auto archIt = std::find(supportedMetricKeys.begin(), supportedMetricKeys.end(),
+            METRIC_KEY(DEVICE_ARCHITECTURE));
+        if (archIt != supportedMetricKeys.end()) {
+            auto value = plugin.GetMetric(METRIC_KEY(DEVICE_ARCHITECTURE), getMetricConfig);
+            compileConfig[METRIC_KEY(DEVICE_ARCHITECTURE)] = value.as<std::string>();
+        } else {
+            // WA: take device name at least if device does not support DEVICE_ARCHITECTURE metric
+            compileConfig[METRIC_KEY(DEVICE_ARCHITECTURE)] = deviceFamily;
+        }
+
+        NetworkCompilationContext context(network, compileConfig);
+        return context.computeHash();
     }
 
     ExecutableNetwork LoadNetwork(const CNNNetwork& network, const std::string& deviceName,
                                   const std::map<std::string, std::string>& config) override {
-        return LoadNetworkImpl(network, deviceName, config, nullptr);
+
+        auto parsed = parseDeviceNameIntoConfig(deviceName, config);
+        auto plugin = GetCPPPluginByName(parsed._deviceName);
+        bool loadedFromCache = false;
+        ExecutableNetwork res;
+        std::string hash;
+        if (DeviceSupportsImportExport(plugin)) {
+            hash = CalculateNetworkHash(network, parsed._deviceName, plugin, config);
```

Comment:
what if network hash calculation throws exception?

Reply:
looks like we need to pass `parsed._config` instead of `config`?

Reply:
Yes, I'll create more tests to verify that proper config (config or parsed._config) is passed to all functions

Reply:
Tests are created. You're right - it shall be "parsed._config" instead of "config"